### PR TITLE
[Linalg] Expose subview promotion as a declarative pattern

### PR DESCRIPTION
**Pull request by tetuante**
This PR targets issue #295. It exposes the already existing
 subiew promotion pass as a declarative pattern.

### CODE REVIEW

File: test/lib/DeclarativeTransforms/TestLinalgTransformPatterns.td [hunk 1]
```diff
@@ -126,4 +125,26 @@ def : Pat<(IndexedGenericOp:$op $_1, $_2, $_3, $_4, $_5, $_6, $_7, $_8),
...
 
+//===----------------------------------------------------------------------===//
+// Linalg subview operands promotion.
+//===----------------------------------------------------------------------===//
+def : Pat<(MatmulOp:$op $A, $B, $C),
+          (LinalgOpPromoteSubviews<"MatmulOp"> $op),
+          [(Constraint<HasOperandsOfType<"SubViewOp">> $op),
+          (Constraint<HasLinalgTransformMarker<"_promote_views_">> $op)]>;
+
+def : Pat<(MatvecOp:$op $A, $b, $c),
```
Comment (nicolasvasilache) [inline test/lib/DeclarativeTransforms/TestLinalgTransformPatterns.td]: These patterns are not used, could we drop them?

File: test/Dialect/Linalg/transform-patterns.mlir [hunk 2]
```diff
@@ -315,3 +315,53 @@ func @matmul_perm(%A: memref<?x?xf32, offset: ?, strides: [?, 1]>,
...
+  %c0 = constant 0 : index
+  %c1 = constant 1 : index
+  %0 = dim %arg0, 0 : memref<?x?xf32, offset: ?, strides: [?, 1]>
+  %1 = dim %arg0, 1 : memref<?x?xf32, offset: ?, strides: [?, 1]>
+  %2 = dim %arg1, 1 : memref<?x?xf32, offset: ?, strides: [?, 1]>
+  loop.for %arg3 = %c0 to %0 step %c2000 {
+    loop.for %arg4 = %c0 to %2 step %c3000 {
+      loop.for %arg5 = %c0 to %1 step %c4000 {
+        %3 = std.subview %arg0[%arg3, %arg5][%c2000, %c4000][%c1, %c1] :
+             memref<?x?xf32, offset: ?, strides: [?, 1]> to memref<?x?xf32, offset: ?, strides: [?, ?]>
```
Comment (nicolasvasilache) [inline test/Dialect/Linalg/transform-patterns.mlir]: The subview result type can keep the same strides as the original type.

File: include/mlir/Dialect/Linalg/Transforms/LinalgTransformPatterns.td [hunk 3]
```diff
@@ -43,6 +43,13 @@ class AffineMapDomainHasDim<int n> : CPred<[{
   $0.getAttrOfType<ArrayAttr>(getIndexingMapsAttrName()).getValue()[0].
   cast<AffineMapAttr>().getValue().getNumDims() ==}] # n # [{}]>;
 
+class HasOperandsOfType<string type>: CPred<[{
+    std::any_of($0.getOperands().begin(), $0.getOperands().end(),
```
Comment (River707) [inline include/mlir/Dialect/Linalg/Transforms/LinalgTransformPatterns.td]: llvm::any_of

File: include/mlir/Dialect/Linalg/Transforms/LinalgTransforms.h [hunk 4]
```diff
@@ -95,6 +95,10 @@ LogicalResult vectorizeGenericOp(PatternRewriter &rewriter, Operation *op);
 LogicalResult permuteGenericLinalgOp(PatternRewriter &rewriter, Operation *op,
                                      ArrayRef<unsigned> permutation,
                                      StringRef linalgMarker);
+
+// Promote std.subviews feeding linalg operations
```
Comment (River707) [inline include/mlir/Dialect/Linalg/Transforms/LinalgTransforms.h]: // -> ///

File: include/mlir/Dialect/Linalg/Utils/Utils.h [hunk 5]
```diff
@@ -23,6 +23,10 @@
 #include "mlir/Dialect/StandardOps/Ops.h"
 #include "mlir/EDSC/Helpers.h"
 
+#include "llvm/ADT/SetVector.h"
+
+using llvm::SetVector;
```
Comment (River707) [inline include/mlir/Dialect/Linalg/Utils/Utils.h]: Don't do this, don't add using directives in a header. Use it in a fully resolved context, or add this to LLVM.h

File: lib/Dialect/Linalg/Transforms/LinalgTransforms.cpp [hunk 6]
```diff
@@ -230,3 +230,19 @@ mlir::linalg::permuteGenericLinalgOp(PatternRewriter &rewriter, Operation *op,
...
+LogicalResult mlir::linalg::linalgOpPromoteSubviews(PatternRewriter &rewriter,
+                                                    Operation *op) {
+  LinalgOp linOp = dyn_cast<LinalgOp>(op);
+  SetVector<Value *> subViews;
+  for (auto it : linOp.getInputsAndOutputs())
+    if (auto sv = dyn_cast_or_null<SubViewOp>(it->getDefiningOp()))
+      subViews.insert(sv);
+  if (!subViews.empty()) {
+    auto resOp = promoteSubViewOperands(rewriter, linOp, subViews);
+    if (!resOp)
```
Comment (River707) [inline lib/Dialect/Linalg/Transforms/LinalgTransforms.cpp]: return success(resOp); ?

File: lib/Dialect/Linalg/Transforms/Promotion.cpp [hunk 7]
```diff
@@ -189,20 +189,22 @@ static void promoteSubViewOperands(LinalgOp op, SetVector<Value *> subViews,
...
+        res.getIndexOfOutput(viewAndPartialLocalView.first).hasValue();
     if (isOutput)
       copy(viewAndPartialLocalView.second, viewAndPartialLocalView.first);
   }
 
   // 4. Dealloc local buffers.
   for (const auto &pi : promotedBufferAndViews)
     dealloc(pi.buffer);
+
+  return res; // std::move?
```
Comment (River707) [inline lib/Dialect/Linalg/Transforms/Promotion.cpp]: Drop the comment here.

File: test/Dialect/Linalg/transform-patterns.mlir (hunk 2 above)
Comment (tetuante) [reply test/Dialect/Linalg/transform-patterns.mlir]: When we tried that (`strides : [?,1]` for the output of the  `subview`), I got the error:
`transform-patterns.mlir:333:14: error: expected result type to have dynamic strides` 
Am I missing something ?

Comment (MaheshRavishankar) [reply test/Dialect/Linalg/transform-patterns.mlir]: THe current specification of subview op doesnt allow mixing and matching static and dynamic strides. Either they should be all static or all dynamic.  So you cannot have [?, ?].
In this case even though the stride operands are constants, the base memref has dynamic stride. So the subview op result must have all strides dynamic.
To fix this, the specification of subview op has to change to allow for this more general representation.

File: lib/Dialect/Linalg/Transforms/Promotion.cpp (hunk 7 above)
Comment (tetuante) [reply lib/Dialect/Linalg/Transforms/Promotion.cpp]: Done. Thanks

File: lib/Dialect/Linalg/Transforms/LinalgTransforms.cpp (hunk 6 above)
Comment (tetuante) [reply lib/Dialect/Linalg/Transforms/LinalgTransforms.cpp]: Much better.Thanks

File: include/mlir/Dialect/Linalg/Transforms/LinalgTransforms.h (hunk 4 above)
Comment (tetuante) [reply include/mlir/Dialect/Linalg/Transforms/LinalgTransforms.h]: Done

File: include/mlir/Dialect/Linalg/Transforms/LinalgTransformPatterns.td (hunk 3 above)
Comment (tetuante) [reply include/mlir/Dialect/Linalg/Transforms/LinalgTransformPatterns.td]: Much better. Thank you.

File: include/mlir/Dialect/Linalg/Utils/Utils.h (hunk 5 above)
Comment (tetuante) [reply include/mlir/Dialect/Linalg/Utils/Utils.h]: Oooops! Done

File: test/lib/DeclarativeTransforms/TestLinalgTransformPatterns.td (hunk 1 above)
Comment (tetuante) [reply test/lib/DeclarativeTransforms/TestLinalgTransformPatterns.td]: Is it worth keeping them and adding some tests for them?

File: test/Dialect/Linalg/transform-patterns.mlir (hunk 2 above)
Comment (nicolasvasilache) [reply test/Dialect/Linalg/transform-patterns.mlir]: ah yes, forgot about that :) thanks!

File: test/lib/DeclarativeTransforms/TestLinalgTransformPatterns.td (hunk 1 above)
Comment (nicolasvasilache) [reply test/lib/DeclarativeTransforms/TestLinalgTransformPatterns.td]: I wouldn't think so, it's very easy to copy-paste the MatmulOp and adapt.

Comment (tetuante) [reply test/lib/DeclarativeTransforms/TestLinalgTransformPatterns.td]: OK. Done.

**Comment by nicolasvasilache** [comment]
@River707 I believe your comments have been addressed.

File: lib/Dialect/Linalg/Transforms/Promotion.cpp [hunk 8]
```diff
@@ -189,20 +189,22 @@ static void promoteSubViewOperands(LinalgOp op, SetVector<Value *> subViews,
...
   opViews.append(operands.begin(), operands.end());
-  op.clone(b, op.getLoc(), opViews);
+  LinalgOp res = op.clone(b, op.getLoc(), opViews);
 
   // 3. Emit write-back for the promoted output views: copy the partial view.
   for (auto viewAndPartialLocalView : writebackViews) {
     // Note: use the old op to determine whether the operand view is an output.
     bool isOutput =
-        op.getIndexOfOutput(viewAndPartialLocalView.first).hasValue();
+        res.getIndexOfOutput(viewAndPartialLocalView.first).hasValue();
```
Comment (nicolasvasilache) [inline lib/Dialect/Linalg/Transforms/Promotion.cpp]: This is a bug that prevent writebacks from happening.
As the comment right above says, you want to check the old op not the newly cloned one.
I'll make it into a WARNING.

Fixing internally for submitting.

**Pull request closed without merging**